### PR TITLE
HPCC-16546 Allow statistics to be filtered by value

### PR DESCRIPTION
**Pull request by ghalliday**
Signed-off-by: Gavin Halliday <gavin.halliday@lexisnexis.com>

### CODE REVIEW

**Comment by ghalliday** [comment]
@richardkchapman this is a bit late in the day, but it makes it much easier to search for slow graphs etc. - and allowing that search to be done from esp rather than the client (in daliadmin).
@afishbeck please can you check the esp changes are correct
@jakesmith please review

If it is too much for 6.2 then I will retarget, and the 6.4 daliadmin could still be used against a 6.x system.

File: esp/scm/ws_workunits.ecm [hunk 1]
```diff
@@ -1718,6 +1719,10 @@ ESPrequest [nil_remove] WUGetStatsRequest
     unsigned MaxScopeDepth;
     boolean IncludeGraphs;
     boolean CreateDescriptions;
+    [min_ver("1.62")] int64 MinValue; // Should be uint64 but esdl does not support it.
+    [min_ver("1.62")] int64 MaxValue;
+    [min_ver("1.62")] string Filter; // arbitrary text filter
+    //MORE: Some indication of which fields need to be returned - to reduce the amount of data returned.  Is there a general esp mechanism?
```
Comment (afishbeck) [inline esp/scm/ws_workunits.ecm]: I don't think there isn't a general mechanism that is relevant here.  There is something more relevant to interface management and versioning, but not run time field selection.

File: esp/scm/ws_workunits.ecm [hunk 2]
```diff
@@ -1718,6 +1719,10 @@ ESPrequest [nil_remove] WUGetStatsRequest
     unsigned MaxScopeDepth;
     boolean IncludeGraphs;
     boolean CreateDescriptions;
+    [min_ver("1.62")] int64 MinValue; // Should be uint64 but esdl does not support it.
```
Comment (afishbeck) [inline esp/scm/ws_workunits.ecm]: Could open a Jira for adding uint64 if you think important enough.

**Review by afishbeck — COMMENTED** [review]
ESP changes look good.

File: dali/daliadmin/daliadmin.cpp [hunk 3]
```diff
@@ -2591,7 +2587,7 @@ static void dumpStats(IConstWorkUnit * workunit, const char * creatorTypeText, c
                 xml.append(count);
             xml.append(",");
             if (max)
-                xml.append(value);
+                xml.append(max);
```
Comment (jakesmith) [inline dali/daliadmin/daliadmin.cpp]: @ghalliday - this was a incidental bug fix?

File: ecl/eclplus/DumpHelper.cpp [hunk 4]
```diff
@@ -17,6 +17,23 @@
...
+{
+    if (excep && excep->ordinality())
+    {
+        unsigned i = 0;
+        while (i < excep->ordinality())
+        {
+            StringBuffer msg;
+            excep->item(i).errorMessage(msg);
+            unsigned code = excep->item(i).errorCode();
+            printf("<Error><code>%d</code><message>%s</message></Error>\n", code, msg.str());
```
Comment (jakesmith) [inline ecl/eclplus/DumpHelper.cpp]: Should this be to stderr?

File: ecl/eclplus/DumpHelper.cpp [hunk 5]
```diff
@@ -147,3 +147,68 @@ bool GraphHelper::doit(FILE * fp)
...
+    if (wuid)
+        req->setWUID(wuid);
+    else
+        req->setWUID("*");
+
+    if (filter)
+        req->setFilter(filter);
+
+    Owned<IClientWUGetStatsResponse> resp = wuclient->WUGetStats(req);
+    const IMultiException* excep = &resp->getExceptions();
```
Comment (jakesmith) [inline ecl/eclplus/DumpHelper.cpp]: Probably just naming convention of response? but it looks like it leaks.

**Comment by jakesmith** [comment]
@ghalliday - couple of minor comment/questions.

File: dali/daliadmin/daliadmin.cpp (hunk 3 above)
Comment (ghalliday) [inline dali/daliadmin/daliadmin.cpp]: yes

File: ecl/eclplus/DumpHelper.cpp (hunk 4 above)
Comment (ghalliday) [inline ecl/eclplus/DumpHelper.cpp]: That would be more sensible, but none of the existing code did (this was extracting existing code into a function)

File: ecl/eclplus/DumpHelper.cpp (hunk 5 above)
Comment (ghalliday) [inline ecl/eclplus/DumpHelper.cpp]: Yes it looks like the esp created interfaces are called getXXX but do not link their return value

**Comment by ghalliday** [comment]
@richardkchapman please review - especially whether it should go in 6.2

**Pull request merged**